env/io_posix.cc

```diff
@@ -56,6 +56,65 @@ int Fadvise(int fd, off_t offset, size_t len, int advice) {
 #endif
 }
 
+// On MacOS (and probably *BSD), the posix write and pwrite calls do not support
+// buffers larger than 2^31-1 bytes. These two wrappers fix this issue by
+// cutting the buffer in 1GB chunks when having to write buffers larger than
+// 2GB. We use this chunk size to be sure to keep the writes aligned.
+
+bool PosixWriteWrapper(int fd, const void* buf, size_t nbyte) {
+  constexpr size_t LIMIT_2GB = 1UL << 31;
+  constexpr size_t LIMIT_1GB = 1UL << 30;
+
+  const char* src = reinterpret_cast<const char*>(buf);
+  size_t left = nbyte;
+
+  while (left != 0) {
+    size_t bytes_to_write = left;
+    if (bytes_to_write >= LIMIT_2GB) {
+      bytes_to_write = LIMIT_1GB;
+    }
+    ssize_t done = write(fd, src, bytes_to_write);
+    if (done < 0) {
+      if (errno == EINTR) {
+        continue;
+      }
+      return false;
+    }
+    left -= done;
+    src += done;
+  }
+  return true;
+}
+
+bool PosixPositionedWriteWrapper(int fd, const void* buf, size_t nbyte,
+                                 off_t offset) {
+  constexpr size_t LIMIT_2GB = 1UL << 31;
+  constexpr size_t LIMIT_1GB = 1UL << 30;
+
+  const char* src = reinterpret_cast<const char*>(buf);
+  size_t left = nbyte;
+
+  while (left != 0) {
+    size_t bytes_to_write = left;
+    if (bytes_to_write >= LIMIT_2GB) {
+      bytes_to_write = LIMIT_1GB;
+    }
+
+    ssize_t done = pwrite(fd, src, bytes_to_write, offset);
+    if (done < 0) {
+      if (errno == EINTR) {
+        continue;
+      }
+      return false;
+    }
+    left -= done;
+    offset += done;
+    src += done;
+  }
+
+  return true;
+}
```

Comment:
It's a good idea to put the two functions into anonymous namespace if possible.

Reply:
This is marked as resolved, but I still didn't see the functions in anonymous namespace. Did I miss anything?

Reply:
My blame: I did something stupid with git and forgot to commit the change. Should be done by now.